scene/resources/curve.cpp

```diff
@@ -1496,76 +1499,138 @@ void Curve3D::_bake() const {
+		baked_up_vector_cache.resize(0);
+		return;
+	}
 
-	Vector3 prev_sideways = Vector3(1, 0, 0);
-	Vector3 prev_up = Vector3(0, 1, 0);
-	Vector3 prev_forward = Vector3(0, 0, 1);
+	// Step 2: Caculate up vectors and whole local reference frames
+	//
+	// see C. Dougan, The Parallel Transport Frame, Game Programming Gems 2 for example
+	// for a discussion about why not Frenet frame.
+	{
+		PackedVector3Array baked_forward_vector_cache;
 
-	for (const Plane &E : pointlist) {
-		w[idx] = E.normal;
-		wt[idx] = E.d;
-		wd[idx] = distlist[idx];
+		baked_up_vector_cache.resize(point_count);
+		baked_forward_vector_cache.resize(point_count);
 
-		if (!up_vector_enabled) {
-			idx++;
-			continue;
+		Vector3 *up_write = baked_up_vector_cache.ptrw();
+		Vector3 *forward_write = baked_forward_vector_cache.ptrw();
+
+		const Vector3 *points_ptr = baked_point_cache.ptr();
+		//const real_t *tilts_ptr = baked_tilt_cache.ptr();
+		//const real_t *dists_ptr = baked_dist_cache.ptr();
+
+		Basis frame; // x is right, y is up, z is forward
+		Basis frame_prev;
+
+		// Set initial frame based on y-up rule
+		{
+			Vector3 up(0, 1, 0);
+			Vector3 forward = (points_ptr[1] - points_ptr[0]).normalized();
+			if (forward.is_equal_approx(Vector3())) {
+				forward = Vector3(1, 0, 0);
+			}
+
+			if (abs(forward.dot(up)) > 1.0 - UNIT_EPSILON) {
+				frame_prev = Basis::looking_at(-forward, up);
+			} else {
+				frame_prev = Basis::looking_at(-forward, Vector3(1, 0, 0));
+			}
+
+			up_write[0] = frame_prev.get_column(1);
+			forward_write[0] = frame_prev.get_column(2);
 		}
 
-		forward = idx > 0 ? (w[idx] - w[idx - 1]).normalized() : prev_forward;
+		// Calculate Parallel Transport Frame
+		for (int idx = 1; idx < point_count; idx++) {
+			Vector3 forward = (points_ptr[idx] - points_ptr[idx - 1]).normalized();
+			if (forward.is_equal_approx(Vector3())) {
+				forward = frame_prev.get_column(2);
+			}
 
-		real_t y_dot = prev_up.dot(forward);
+			Basis rotate;
+			rotate.rotate_to_align(frame_prev.get_column(2), forward);
+			frame = rotate * frame_prev;
+			frame.orthonormalize(); // guard against float error accumulation
 
-		if (y_dot > (1.0f - CMP_EPSILON)) {
-			sideways = prev_sideways;
-			up = -prev_forward;
-		} else if (y_dot < -(1.0f - CMP_EPSILON)) {
-			sideways = prev_sideways;
-			up = prev_forward;
-		} else {
-			sideways = prev_up.cross(forward).normalized();
-			up = forward.cross(sideways).normalized();
+			up_write[idx] = frame.get_column(1);
+			forward_write[idx] = frame.get_column(2);
+
+			frame_prev = frame;
 		}
 
-		if (idx == 1) {
-			up_write[0] = up;
+		// loop smoothing only applies when the curve is a loop
+		bool is_loop = true;
+		// a loop: two ends meet, and share forward directions
+		{
+			if (!points_ptr[0].is_equal_approx(points_ptr[point_count - 1])) {
+				is_loop = false;
+			}
+
+			real_t dot = forward_write[0].dot(forward_write[point_count - 1]);
+			if (dot < 1.0 - UNIT_EPSILON) {
+				is_loop = false;
+			}
```

Comment:
Is this check necessary? Would it help to prevent the model upside down in a case like a folded diamond shape? Also, this check prevents smoothing Depends on the value of `bake_interval`.

I assume that this checks exist to address this latter issue. It would be better if the twist could be independent of `bake_interval`.

But even if not, the current smoothing process is better than nothing. So I suggest making a warning with `WARN_PRINT()` that "Smoothing is done since curve path is closed, but it may not work well because of bake interval is not suitable or sharp angle between end point and start point." instead of not performing smoothing by checking. It will warn the user only once for each curve bake. This will give someone else a chance to improve the smoothing later.

Reply:
> help to prevent the model upside down in a case like a folded diamond shape?

I assume you mean the model roll 180 degree around the tangent in a very short distance at narrow corner on the curve? Then yes.

This is an intrinsic problem of the Frenet frame. Because at two side of the corner, the Tangent flipped, but Binormal didn't change, so the Normal(up vector) have to flip. See my drawing.

![image](https://user-images.githubusercontent.com/6020486/200214674-5ab5f750-eb4c-40cb-a481-00f2042f56d5.png)


So my implementation use another technique called [Transport Frame](https://gitea.yiem.net/QianMo/Real-Time-Rendering-4th-Bibliography-Collection/raw/branch/main/Chapter%201-24/[0374]%20[Game%20Programming%20Gems%202%202001]%20The%20Parallel%20Transport%20Frame.pdf), which guarantee the up vector change smoothly if the curve do not have sharp corner(z shape).

> Is this check necessary? 

I think yes. It checkes if the loop joint is smooth(tangents aligned at two ends).

Transport Frame introduce a new problem. For closed loop, the Frenet Frame guarantee the up vector well defined at the joint point, as long as the joint is smooth(the tangents of two ends align). The Transport Frame do not have this property, and the up vector will be different for the begin and end of the loop, which means a sudden change across the joint point.

![image](https://user-images.githubusercontent.com/6020486/200214752-1577be7d-6fea-415f-8edd-f0e4fee7e20b.png)

This  problem can be fixed by making the up vectors at the begin and end align, which require rolling up vectors gradually along the curve.  I think this fix is only meaningful if the joint is smooth(tangent align).

![image](https://user-images.githubusercontent.com/6020486/200214942-97bc7703-4a8c-4b62-ad7e-3d329903af44.png)

>  Also, this check prevents smoothing Depends on the value of bake_interval

I don't quite follow. Can you elaborate? I think this depends on the whether the tangents at two ends of the curve align.

Reply:
When `bake_interval` is 0.2 which is the default value, there are often cases where the joint is smooth but the curve is a little steep (e.g. bowling pin shape) there are cases where the angle is different when progress ratio is 0 and 1. Setting `bake_interval` to 0.01 will It improves, but it is too heavy.

Reply:
That is strange. Is there example file? I'd like to investigate.

Reply:
Bake interval 0.2:
- progress 0:
![image](https://user-images.githubusercontent.com/61938263/200263140-e1cfc323-2d4b-4598-8e93-067e25085b23.png)
- progress 1:
![image](https://user-images.githubusercontent.com/61938263/200263199-3ce069a4-2160-4458-863c-4fe38db31282.png)

Bake interval 0.01:
- progress 0:
![image](https://user-images.githubusercontent.com/61938263/200263396-601c03e9-07ca-4b9f-bfd6-1e0971ba772b.png)
- progress 1:
![image](https://user-images.githubusercontent.com/61938263/200263462-971e2287-34fa-4c79-891e-33ff9772fee5.png)

[path.zip](https://github.com/godotengine/godot/files/9949554/path.zip)

Reply:
Tangent alignment check is too tight. Loosen it solve the problem